fix: overwrite when Pact has multiple interactions

asteffey opened the pull request:
## Problem to Fix

Fixes a problem where only the last interaction defined for
a Pact was recorded if pact.writer.overwrite is set to
true.

### Steps to reproduce

- Set `pact.writer.overwrite` system property to true
- Include two or more `@Pact` annotated methods to define different interactions for the same consumer and provider combination

### What is the current *bug* behavior?

Pact will only include the interaction defined in the last `@Pact` method.

### What is the expected *correct* behavior?

Pact contract should include all interactions defined in a single class, even when `pact.writer.overwrite` is `true`.

## Proposed Solution

Write Pacts at end during `afterAll` instead of after testing each interaction in `afterTestExecution`.  Saved each Pact in `store["pactsToWrite"]` and used `Pact::mergeInteractions` to combine different interactions.

### Unintended Consequences

This broke the `checkPactFile` test in `ProviderStateInjectedPactTest`, because the Pact file is now being written after this occurs.  To work-around, had to remove method and moved each of the assertions within the test elsewhere.  Moved the generators assertions to `articles` method within the same file.  Moved the version assertion to the new test in `PactConsumerTestExtSpec`.

asteffey commented:
@jstandard, here's another bug fix for review.

jstandard commented on consumer/pact-jvm-consumer-junit5/src/main/kotlin/au/com/dius/pact/consumer/junit5/PactConsumerTestExt.kt:
```diff
@@ -308,35 +310,38 @@ class PactConsumerTestExt : Extension, BeforeTestExecutionCallback, BeforeAllCal
       val store = context.getStore(NAMESPACE)
       val providerInfo = store["providerInfo"] as ProviderInfo
-      val pactDirectory = lookupPactDirectory(context)
       if (providerInfo.providerType != ProviderType.ASYNCH) {
```
I realize the original code had the current order, but for clarity, I might switch the logic putting ProviderType.ASYNCH first...

```
override fun afterTestExecution(context: ExtensionContext) {
    if (!context.executionException.isPresent) {
      val store = context.getStore(NAMESPACE)
      val providerInfo = store["providerInfo"] as ProviderInfo
      if (providerInfo.providerType == ProviderType.ASYNCH) {
        storePactForWrite(store)
      } else {
        val mockServer = store["mockServer"] as JUnit5MockServerSupport
        Thread.sleep(100) // give the mock server some time to have consistent state
        mockServer.close()
        val result = mockServer.validateMockServerState(null)
        if (result is PactVerificationResult.Ok) {
          storePactForWrite(store)
        } else {
          JUnitTestSupport.validateMockServerResult(result)
        }
      }
    }
}
```

asteffey replied:
Change made.

jstandard reviewed (commented):
In general, looks good.  I only had one minor comment.

asteffey committed:
fix: Don't overwrite Pact with multiple interactions

Fix a problem where only the last interaction defined for
a Pact was written if pact.writer.overwrite is set to
true. Write Pacts at end in afterAll instead of each
afterTestExecution.
